### ISSUE TITLE
Broadcast Preview to Browser

### ISSUE THREAD

**Issue by NoahCardoza** [atom-community/markdown-preview-plus#411]
Ok so basically, I was looking into [atom-broadcast](https://github.com/eqot/atom-broadcast) and after installing it, it wouldn't work with `markdown-preview-plus`. But I was wondering if I could somehow merge both those repositories.

I understand this is probably not a feature which you would like to integrate so I am just asking if anyone could point me in the right direction. I know there is a way to save as HTML (which I use a lot 👍 thanks!)  could I perhaps link that to the broadcast output and serve that as the update?

I hope this makes sense...

If this is overly complicated and there is a simpler way please, do tell. The whole reason for this is I am teaching a class and I'd like to be able to share my slides (a.k.a. my md doc) to the class over the local network.

Thanks in advance!

**Comment by lierdakil** [comment]
This issue seems to be in the wrong place, frankly. atom-broadcast needs to support MPP directly, probably somewhere [around here](https://github.com/eqot/atom-broadcast/blob/f41dc0bf9ef9adcbb23281451be963ff52ddc44c/lib/broadcast-target.coffee)

I'll gladly work with package authors to establish some sort of API if we need to.

> The whole reason for this is I am teaching a class and I'd like to be able to share my slides (a.k.a. my md doc) to the class over the local network.

I'm... not entirely sure how atom-broadcast helps you in this regard in particular? I mean, serving a presentation over the LAN using a simple http server is reasonably easy anyway (see e.g. https://github.com/hakimel/reveal.js/#multiplexing, which might or might not be overkill for your application in particular). Unless you need to edit that on-the-fly. If you do, you could probably get away with a live-reload dev server of some description (e.g. http://tapiov.net/live-server/). If you don't `python -m SimpleHTTPServer` or something similar might just work as well.

**Comment by NoahCardoza** [comment]
I had looked into reveal.js but I was hoping to edit live. I really like the hot reloading idea, I think I'll look into that. I just hope I can find a nice markdown plugin. I especially like the way markdown-preview-plus handles the code formatting colors since it is the same as my ide, but I guess I could just copy the CSS rules from the HTML file and maybe apply those to another NPM package.

Thanks for the reply!

**Comment by lierdakil** [comment]
> I just hope I can find a nice markdown plugin.

You don't really need to though. Here's a workflow sketch:

1. Export HTML from MPP to some location (e.g. `server/index.html`)
2. Run live-update dev server that would serve HTML from (1)
3. Make changes to Markdown.
4. Export HTML again to the same location as in (1). live-update server should notice the change and update clients accordingly.
5. Rinse and repeat (3)-(5) until satisfied

Yes, this has additional step (4) you have to do. But that is somewhat easier to amend. For example, you could write a 20-odd line script to automatically save the last updated markdown preview to a predefined location. Something to the tune of
```coffee
atom.workspace.observePaneItems item =>
  if item.onDidChangeMarkdown
    item.onDidChangeMarkdown => item.saveAs '/path/to/a/predefined/location.html'
```
(in coffeescript, pretty much the same in pure JS, but with more parentheses -- bear in mind I didn't check if this works)

It is, of course, a bodge. But a good bodge is almost as good as a proper solution.

**Comment by lierdakil** [comment]
> But if you found a way to make it work that would be amazing!
As I said, support for MPP *has* to be explicitly implemented in atom-broadcast. No way around that. Current iteration of MPP is *very* different internally from built-in markdown-preview. 

Also, there's currently no API in MPP to (easily) get HTML source -- but that's amendable, it just needs some discussion. In particular, what should be exported from MPP, complete HTML with styles and scripts, or just body, or otherwise? Can atom-broadcast handle MathJax, or should MPP convert math to SVG? Etc.

/cc @eqot -- does this sound like something you'd be interested in? For context, talking about markdown-preview-plus support in atom-broadcast.

**Comment by NoahCardoza** [comment]
I gave you suggestion a try (Exporting HTML from MPP, not the automatic version though) and noticed that indentation inside of the code blocks is not retained, which is especially an issue when demoing Python code.

Not sure if this is a known issue or I did something wrong. Should I create a new issue for it?

**Comment by lierdakil** [comment]
Not a known issue, some style leakage into exported HTML occurred. Thanks for reporting. Should be fixed in v3.3.2
